web/src/main.tsx

```diff
@@ -63,6 +73,116 @@ const queryClient = new QueryClient({
 
 refetchDataOnHourChange(queryClient);
 
+export function ValidZoneIdGuardWrapper({ children }: { children: JSX.Element }) {
+  const [searchParameters] = useSearchParams();
+  const { zoneId, timeAverage } = useParams();
+  if (!zoneId) {
+    return <Navigate to="/map/hourly" replace />;
+  }
+  const upperCaseZoneId = zoneId.toUpperCase();
+  if (zoneId !== upperCaseZoneId) {
+    return (
+      <Navigate
+        to={`/zone/${upperCaseZoneId}/${timeAverage}?${searchParameters}`}
+        replace
+      />
+    );
+  }
+
+  // Handle legacy Australia zone names
+  if (upperCaseZoneId.startsWith('AUS')) {
+    return (
+      <Navigate
+        to={`/zone/${zoneId.replace('AUS', 'AU')}/${timeAverage}?${searchParameters}`}
+        replace
+      />
+    );
+  }
+
+  // Only allow valid zone ids
+  // TODO: This should redirect to a 404 page specifically for zones
+  if (!zoneExists(upperCaseZoneId)) {
+    return <Navigate to="/map/hourly" replace />;
+  }
+
+  return children;
+}
+
+function HandleLegacyRoutes() {
+  const [searchParameters] = useSearchParams();
```

Comment:
Consider treating `searchParameters` as immutable - maybe make a copy of the map and modify the copy, which can be used for the [`to` prop on line 126](https://github.com/electricitymaps/electricitymaps-contrib/pull/7361/files#diff-20da45cb8443d5036aea448ba7f0379ca304c43e1136ca82d9d43d731a37311eR124)